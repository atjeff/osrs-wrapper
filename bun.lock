{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "osrs-wrapper",
      "devDependencies": {
        "@devoxa/prettier-config": "^2.0.3",
        "@tsconfig/node22": "^22.0.0",
        "@types/node": "^22.13.10",
        "bun-types": "^1.2.5",
        "prettier": "^3.5.3",
        "typescript": "^5.8.2",
      },
    },
  },
  "packages": {
    "@devoxa/prettier-config": ["@devoxa/prettier-config@2.0.3", "", { "dependencies": { "prettier-plugin-organize-imports": "3.2.4" }, "peerDependencies": { "prettier": ">=2.4", "typescript": ">=2.9" } }, "sha512-g1ajCzaL70dW/RlWTtWRxslOrLh5Gir2orsdqzN1HjJFQhJ1AmoBn7lQgepiDfvAdjG+fMMfLuPm2TUzEvO4cg=="],

    "@tsconfig/node22": ["@tsconfig/node22@22.0.0", "", {}, "sha512-twLQ77zevtxobBOD4ToAtVmuYrpeYUh3qh+TEp+08IWhpsrIflVHqQ1F1CiPxQGL7doCdBIOOCF+1Tm833faNg=="],

    "@types/node": ["@types/node@22.13.10", "", { "dependencies": { "undici-types": "~6.20.0" } }, "sha512-I6LPUvlRH+O6VRUqYOcMudhaIdUVWfsjnZavnsraHvpBwaEyMN29ry+0UVJhImYL16xsscu0aske3yA+uPOWfw=="],

    "@types/ws": ["@types/ws@8.5.14", "", { "dependencies": { "@types/node": "*" } }, "sha512-bd/YFLW+URhBzMXurx7lWByOu+xzU9+kb3RboOteXYDfW+tr+JZa99OyNmPINEGB/ahzKrEuc8rcv4gnpJmxTw=="],

    "bun-types": ["bun-types@1.2.5", "", { "dependencies": { "@types/node": "*", "@types/ws": "~8.5.10" } }, "sha512-3oO6LVGGRRKI4kHINx5PIdIgnLRb7l/SprhzqXapmoYkFl5m4j6EvALvbDVuuBFaamB46Ap6HCUxIXNLCGy+tg=="],

    "prettier": ["prettier@3.5.3", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-QQtaxnoDJeAkDvDKWCLiwIXkTgRhwYDEQCghU9Z6q03iyek/rxRh/2lC3HB7P8sWT2xC/y5JDctPLBIGzHKbhw=="],

    "prettier-plugin-organize-imports": ["prettier-plugin-organize-imports@3.2.4", "", { "peerDependencies": { "@volar/vue-language-plugin-pug": "^1.0.4", "@volar/vue-typescript": "^1.0.4", "prettier": ">=2.0", "typescript": ">=2.9" }, "optionalPeers": ["@volar/vue-language-plugin-pug", "@volar/vue-typescript"] }, "sha512-6m8WBhIp0dfwu0SkgfOxJqh+HpdyfqSSLfKKRZSFbDuEQXDDndb8fTpRWkUrX/uBenkex3MgnVk0J3b3Y5byog=="],

    "typescript": ["typescript@5.8.2", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-aJn6wq13/afZp/jT9QZmwEjDqqvSGp1VT5GVg+f/t6/oVyrgXM6BY1h9BRh/O5p3PlUPAe+WuiEZOmb/49RqoQ=="],

    "undici-types": ["undici-types@6.20.0", "", {}, "sha512-Ny6QZ2Nju20vw1SRHe3d9jVu6gJ+4e3+MMpqu7pqE5HT6WsTSlce++GQmK5UXS8mzV8DSYHrQH+Xrf2jVcuKNg=="],

    "@types/ws/@types/node": ["@types/node@20.17.24", "", { "dependencies": { "undici-types": "~6.19.2" } }, "sha512-d7fGCyB96w9BnWQrOsJtpyiSaBcAYYr75bnK6ZRjDbql2cGLj/3GsL5OYmLPNq76l7Gf2q4Rv9J2o6h5CrD9sA=="],

    "bun-types/@types/node": ["@types/node@20.17.24", "", { "dependencies": { "undici-types": "~6.19.2" } }, "sha512-d7fGCyB96w9BnWQrOsJtpyiSaBcAYYr75bnK6ZRjDbql2cGLj/3GsL5OYmLPNq76l7Gf2q4Rv9J2o6h5CrD9sA=="],

    "@types/ws/@types/node/undici-types": ["undici-types@6.19.8", "", {}, "sha512-ve2KP6f/JnbPBFyobGHuerC9g1FYGn/F8n1LWTwNxCEzd6IfqTwUQcNXgEtmmQ6DlRrC1hrSrBnCZPokRrDHjw=="],

    "bun-types/@types/node/undici-types": ["undici-types@6.19.8", "", {}, "sha512-ve2KP6f/JnbPBFyobGHuerC9g1FYGn/F8n1LWTwNxCEzd6IfqTwUQcNXgEtmmQ6DlRrC1hrSrBnCZPokRrDHjw=="],
  }
}
