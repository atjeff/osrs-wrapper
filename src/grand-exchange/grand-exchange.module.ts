import { request } from 'gaxios';
import { ExchangeGraph, Item, ItemDetailResponse } from './item.model';

/**
 * Gets Grand Exchange items for specified Item ID.
 *
 * @export
 * @param {number} itemId
 * @returns {Promise<Item>}
 */
export async function getExchangeStats(itemId: number): Promise<Item> {
    const {
        data: { item },
    } = await request<ItemDetailResponse>({
        method: 'GET',
        url: 'http://services.runescape.com/m=itemdb_oldschool/api/catalogue/detail.json',
        params: { item: itemId },
    });

    return item;
}

/**
 * Returns Grand Exchange graph points for specified Item ID.
 *
 * @export
 * @param {number} itemId
 * @returns {Promise<ExchangeGraph>}
 */
export async function getExchangeTrendGraph(itemId: number): Promise<ExchangeGraph> {
    const { data } = await request<ExchangeGraph>({
        method: 'GET',
        url: `http://services.runescape.com/m=itemdb_oldschool/api/graph/${itemId}.json`,
    });

    return data;
}
