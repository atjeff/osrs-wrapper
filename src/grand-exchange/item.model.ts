export interface ItemDetailResponse {
  item: Item
}

export interface Item {
  id: number
  name: string
  description: string
  icon: string
  icon_large: string
  type: string
  typeIcon: string
  members: string
  current: TrendPrice
  today: TrendPrice
  day30: TrendChange
  day90: TrendChange
  day180: TrendChange
}

export interface TrendPrice {
  trend: 'positive' | 'negative' | 'neutral'
  price: string
}

export type PriceChange = `${number}%`
export type TrendChange = {
  trend: 'positive' | 'negative' | 'neutral'
  change: PriceChange
}

export interface ExchangeGraph {
  daily: GraphPoints
  average: GraphPoints
}

export type GraphPoints = Record<string, number>
