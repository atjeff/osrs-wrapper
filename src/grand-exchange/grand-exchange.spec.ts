import { afterAll, afterEach, beforeEach, describe, expect, it, spyOn } from 'bun:test'
import { mockExchangeGraph, mockItemResponse } from '../../test/item-response.mock'
import { getExchangeStats, getExchangeTrendGraph } from './grand-exchange.module'

const spyFetch = spyOn(globalThis, 'fetch')

beforeEach(() => {
  spyFetch.mockReset()
})
afterEach(() => spyFetch.mockClear())
afterAll(() => spyFetch.mockRestore())

describe('Grand Exchange', () => {
  const itemId = 4151

  describe('getExchangeStats', () => {
    beforeEach(() => {
      const mockedSuccessfulResponse = {
        ok: true,
        json: () => Promise.resolve(mockItemResponse),
      } as any
      spyFetch.mockResolvedValue(mockedSuccessfulResponse)
    })

    it('should make a GET request', async () => {
      await getExchangeStats(itemId)

      expect(spyFetch.mock.calls[0][0]).toMatchInlineSnapshot(
        `"http://services.runescape.com/m=itemdb_oldschool/api/catalogue/detail.json?item=4151"`
      )
    })

    it('should make a request with the passed in itemId', async () => {
      await getExchangeStats(itemId)

      expect(spyFetch.mock.calls[0][0]).toMatchInlineSnapshot(
        `"http://services.runescape.com/m=itemdb_oldschool/api/catalogue/detail.json?item=4151"`
      )
    })

    it('should return the item key from the response', async () => {
      const output = await getExchangeStats(itemId)

      expect(output).toEqual(mockItemResponse.item)
    })
  })

  describe('getExchangeTrendGraph', () => {
    beforeEach(() => {
      const mockGraphResponse = {
        ok: true,
        json: () => Promise.resolve(mockExchangeGraph),
      } as any
      spyFetch.mockResolvedValue(mockGraphResponse)
    })

    it('should make a GET request', async () => {
      await getExchangeTrendGraph(itemId)

      expect(spyFetch.mock.calls[0][0]).toMatchInlineSnapshot(
        `"http://services.runescape.com/m=itemdb_oldschool/api/graph/4151.json"`
      )
    })

    it('should make a request with the passed in itemId', async () => {
      await getExchangeTrendGraph(itemId)

      expect(spyFetch.mock.calls[0][0]).toMatchInlineSnapshot(
        `"http://services.runescape.com/m=itemdb_oldschool/api/graph/4151.json"`
      )
    })

    it('should return the response', async () => {
      const output = await getExchangeTrendGraph(itemId)

      expect(output).toEqual(mockExchangeGraph)
    })
  })
})
