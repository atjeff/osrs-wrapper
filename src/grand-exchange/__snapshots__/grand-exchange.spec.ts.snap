// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grand Exchange getExchangeStats should make a GET request 1`] = `
{
  "method": "GET",
  "params": {
    "item": 4151,
  },
  "url": "http://services.runescape.com/m=itemdb_oldschool/api/catalogue/detail.json",
}
`;

exports[`Grand Exchange getExchangeStats should make a request with the passed in itemId 1`] = `
{
  "method": "GET",
  "params": {
    "item": 4151,
  },
  "url": "http://services.runescape.com/m=itemdb_oldschool/api/catalogue/detail.json",
}
`;

exports[`Grand Exchange getExchangeTrendGraph should make a GET request 1`] = `
{
  "method": "GET",
  "url": "http://services.runescape.com/m=itemdb_oldschool/api/graph/4151.json",
}
`;

exports[`Grand Exchange getExchangeTrendGraph should make a request with the passed in itemId 1`] = `
{
  "method": "GET",
  "url": "http://services.runescape.com/m=itemdb_oldschool/api/graph/4151.json",
}
`;
